Data/SBV/Core/Symbolic.hs

```diff
@@ -1856,11 +1857,10 @@ uncacheGen getCache (Cached f) st = do
         let rCache = getCache st
         stored <- readIORef rCache
         sn <- f `seq` makeStableName f
-        let h = hashStableName sn
-        case (h `IMap.lookup` stored) >>= (sn `lookup`) of
```

Comment:
Note the second lookup here; which is necessary for correctness. We first do a `IMap` lookup, and then do a regular walk-through the list itself.

I'd be perfectly happy to replace the `IMap` wuth `HMap` and the list here with a faster lookup structure; but we should definitely allow for collisions.

Reply:
Ok, I wasn't sure how common collisions are in practice but this certainly explains the list. Even if we maintain an `O(n)` lookup the benefit is reducing garbage collection pressure. So its less about runtime performance than memory performance.